### PR TITLE
Workflow Execution Timeout Timer

### PR DESCRIPTION
**Pull request by yycptt**
## What changed?

- Add new timer task type for workflow execution timeout & carry over execution timeout timer state across runs in a workflow chain.
- This change depends on https://github.com/temporalio/temporal/pull/5531 so that only one timer task will be created on the standby side as well. 
- New behavior is controlled via a feature flag.

## Why?

- For a workflow chain, we only need one execution timeout timer task for the entire chain. Before this change, each run in the chain will create a run timeout timer. If user set runTimeout == execution timeout (which is the default if only execution timeout is specified!), then the timeout timer for all the runs will fire at the exact same time (because run timeout will be shortened to respect the execution timeout). This can result in a timer burst for a single workflowID. 

## How did you test it?

- Added unit tests
- Functional tests WIP
- Tested locally with cron workflows.

## Potential risks

- In worst case, workflow may not timeout properly.

## Documentation

- N/A, no change on behavior.

## Is hotfix candidate?

- No.

### CODE REVIEW

File: service/history/timer_queue_standby_task_executor.go [hunk 1]
```diff
@@ -426,6 +425,38 @@ func (t *timerQueueStandbyTaskExecutor) executeWorkflowTimeoutTask(
 	)
 }
 
+func (t *timerQueueStandbyTaskExecutor) executeWorkflowExecutionTimeoutTask(
+	ctx context.Context,
+	timerTask *tasks.WorkflowExecutionTimeoutTask,
+) error {
+	actionFn := func(
```
Comment (yiminc) [inline service/history/timer_queue_standby_task_executor.go:432]: could you put some comment here what this is doing

File: service/history/task_executor.go [hunk 2]
```diff
@@ -54,15 +54,33 @@ func getWorkflowExecutionContextForTask(
...
 }
 
 func getWorkflowExecutionContext(
 	ctx context.Context,
 	shardContext shard.Context,
 	workflowCache wcache.Cache,
 	key definition.WorkflowKey,
+	lockPriority workflow.LockPriority,
 ) (workflow.Context, wcache.ReleaseCacheFunc, error) {
+	if key.GetRunID() == "" {
```
[the remark this answers is not in the document]
Comment (yycptt) [reply service/history/task_executor.go:73]: yeah the consistency check in api handler basically does the same thing. I want to refactor after flags like lock current execution and shard ownership assertion are deprecated, that will make things a lot easier.

File: common/persistence/serialization/task_serializer.go [hunk 3]
```diff
@@ -808,35 +813,66 @@ func (s *TaskSerializer) timerUserTaskFromProto(
...
+	}
+}
+
+func (s *TaskSerializer) timerWorkflowExecutionToProto(
+	workflowExecutionTimer *tasks.WorkflowExecutionTimeoutTask,
+) *persistencespb.TimerTaskInfo {
+	return &persistencespb.TimerTaskInfo{
+		NamespaceId:         workflowExecutionTimer.NamespaceID,
+		WorkflowId:          workflowExecutionTimer.WorkflowID,
+		RunId:               "",
```
[the remark this answers is not in the document]
Comment (yycptt) [reply common/persistence/serialization/task_serializer.go:839]: I have comment here: https://github.com/temporalio/temporal/pull/5588/files#diff-9f48fd91d298a8400a6b829050b7abdc26627704a1898930b73cec6cb64e0665R59

Will change serializer to call ^ method instead.

File: service/history/workflow/cache/cache.go [hunk 4]
```diff
@@ -387,3 +392,51 @@ func (c *CacheImpl) validateWorkflowID(
...
+func GetCurrentRunID(
+	ctx context.Context,
+	shardContext shard.Context,
+	workflowCache Cache,
+	shardOwnershipAsserted *bool,
+	namespaceID string,
+	workflowID string,
+	lockPriority workflow.LockPriority,
+) (runID string, retErr error) {
+	if shardContext.GetConfig().EnableAPIGetCurrentRunIDLock() {
```
[the remark this answers is not in the document]
Comment (yycptt) [reply service/history/workflow/cache/cache.go:405]: yes. Will do in a separate PR.

File: service/history/workflow/mutable_state_impl.go [hunk 5]
```diff
@@ -445,6 +445,7 @@ func NewSanitizedMutableState(
 	}
 
 	// sanitize data
```
[the remark this answers is not in the document]
Comment (yycptt) [reply service/history/workflow/mutable_state_impl.go:447]: Done. I also mentioned it [here](https://github.com/temporalio/temporal/pull/5588/files#diff-13ca8d134ba4cff9f3f96095383911aa95f40e06d8afe83d3edaa5f5ef5c304eR177)

File: service/history/workflow/mutable_state_impl.go [hunk 6]
```diff
@@ -457,6 +458,35 @@ func NewSanitizedMutableState(
...
+		namespaceEntry,
+		workflowID,
+		runID,
+		startTime,
+	)
+
+	// carry over necessary fields from current mutable state
+	newMutableState.executionInfo.WorkflowExecutionTimerTaskStatus = currentMutableState.GetExecutionInfo().WorkflowExecutionTimerTaskStatus
+
+	// TODO: carry over other fields like autoResetPoints, previousRunID, firstRunID, etc.
```
[the remark this answers is not in the document]
Comment (yycptt) [reply service/history/workflow/mutable_state_impl.go:484]: We already carry over those fields but in a different place, when adding startEvent (AddWorkflowExecutionStartedEventWithOptions). 

The comment means ideally there should all be carried over here, as not all information need be carried over is part of the started event. I will make the comment more clear

File: service/history/workflow/mutable_state_rebuilder.go [hunk 7]
```diff
@@ -699,15 +701,38 @@ func (b *MutableStateRebuilderImpl) applyNewRunHistory(
...
-	newRunMutableState := NewMutableState(
-		b.shard,
-		b.shard.GetEventsCache(),
-		b.logger,
-		b.mutableState.GetNamespaceEntry(),
-		newExecution.WorkflowId,
-		newExecution.RunId,
-		timestamp.TimeValue(newRunHistory[0].GetEventTime()),
-	)
+	// TODO: replication task should contain enough information to determine whether the new run is part of the same chain
```
[the remark this answers is not in the document]
Comment (yycptt) [reply service/history/workflow/mutable_state_rebuilder.go:704]: TerminateIfRunning will also result in a replication task with new run, but they are not the same chain.

File: service/history/timer_queue_standby_task_executor.go (hunk 1 above)
Comment (yycptt) [reply service/history/timer_queue_standby_task_executor.go:432]: done.

File: service/history/workflow/cache/cache.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (yycptt) [reply service/history/workflow/cache/cache.go:405]: https://github.com/temporalio/temporal/pull/5637

File: service/history/workflow/task_generator.go [hunk 8]
```diff
@@ -133,24 +133,69 @@ func NewTaskGenerator(
...
+	//
+	// Also note the run timeout, if not specified, defaults to execution timeout, so we won't run
+	// into the situation where execution timeout is set but no timeout timer task is generated.
+
+	isFirstRun := executionInfo.FirstExecutionRunId == r.mutableState.GetExecutionState().RunId
+	workflowExecutionExpirationTime := timestamp.TimeValue(
+		executionInfo.WorkflowExecutionExpirationTime,
+	)
+	if workflowExecutionTimeoutTimerEnabled &&
+		!isFirstRun &&
```
Comment (yycptt) [inline service/history/workflow/task_generator.go:169]: cc @yiminc 
This is the improvement we discussed yesterday, essentially don't generate execution timeout timer for the first run.

**Review by yiminc — APPROVED** [review]

**Pull request merged**